src/modules/roc_core/target_musl/roc_core/backtrace.cpp

```diff
@@ -6,15 +6,218 @@
+ */
+void backtrace_symbols()
+{
+	/* To point to the current frame in the call stack.
+	 */
+	unw_cursor_t cursor;
+
+	/* To store the snapshot of the CPU registers.
+	 */
+	unw_context_t context;
+
+	/* To get snapshot of the CPU register.
+	 * unw_getcontext() is signal safe.
+	 */
+	unw_getcontext(&context);
+
+	/* To point to current frame. 
+	 */
+	unw_init_local(&cursor, &context);
+
+	/* Moving to previously called frames & going through each of them.
+	 */
+	int i = 0;
+	while (unw_step(&cursor) > 0) {
+		char b[128];
+		unw_word_t offset, ip;
+		b[0] = '\0';
+		/* Get value stored in instruction register.
+		 * unw_get_reg() is signal safe.
+		 */
+		unw_get_reg(&cursor, UNW_REG_IP, &ip);
+
+		/* Get function name, offset.
+		 * unw_get_proc_name() is signal safe.
+		 */
+		(void) unw_get_proc_name(&cursor, b, sizeof(b), &offset);
+		
+		/* Perform demangling
+		 */
+		char *demangled;
+		int status = -1;
+		if(offset > 0) {
+			char mangled[128];
+			memcpy(mangled, b, 128);
+			demangled = abi::__cxa_demangle(mangled, 0, 0, &status);
+		}
+		/* printing index, function name, cursor, offset, ip
+		 */
+		fprintf(stderr, "#%d : (%s+0x%x) [0x%x]\n", ++i, (status == 0)? demangled : b, (u_int)offset, (u_int)ip);
+	}
+}
+
+/* This must be signal safe.
+ */
+void backtrace_symbols_fd(int fd){
+	/* To point to the current frame in the call stack.
+	 */
+	unw_cursor_t cursor;
+
+	/* To store the snapshot of the CPU registers.
+	 */
+	unw_context_t context;
+
+	/* To get snapshot of the CPU register.
+	 * unw_getcontext() is signal safe.
+	 */
+	unw_getcontext(&context);
+
+	/* To point to current frame.
+	 * unw_init_local() is signal safe. 
+	 */
+	unw_init_local(&cursor, &context);
+
+	/* Moving to previously called frames & going through each of them.
+	 * unw_step() is signal safe.
+	 * buffer = final output that will be written to stderr. 
+	 */
+	u_int index = 0;
+	while (unw_step(&cursor) > 0) {
+		char buffer[200] = "#";
+		index++;
+		char function_name[128];
+		unw_word_t offset, ip;
+		function_name[0] = '\0';
+		/* Get value stored in instruction register
+		 * unw_get_reg() is signal safe.
+		 */
+		unw_get_reg(&cursor, UNW_REG_IP, &ip);
+		/* Get function name & offset
+		 * unw_get_proc_name() is signal safe.
+		 */
+		(void) unw_get_proc_name(&cursor, function_name, 40, &offset);
+		/* Printing index, function name, cursor, offset, ip
+		 * using write(). 
+		 * write() & strcat() are signal safe.
+		 */
+		char index_buffer[10];
+		uint_to_string(index, 10, index_buffer);
+		strcat(buffer, index_buffer);
+		strcat(buffer, ": (");
+		strcat(buffer, function_name);
+		strcat(buffer, "+0x");
+		
+		char offset_buffer[10];
+		uint_to_string((u_int)offset, 16, offset_buffer);
+		strcat(buffer, offset_buffer);
+		strcat(buffer, ") [");
+
+		char address_buffer[16];
+		uint_to_string((u_int)ip, 16, address_buffer);
+		strcat(buffer, address_buffer);
+		strcat(buffer, "]\n");
+
+		write(fd, buffer, 200);
```

Comment:
There are two concerns here:

* we should handle partial writes
* we should use something portable, since write() is not available on all platforms supported by libunwind

For this purpose, I've extracted print_emergency_message() function: 9503e5ea9c324dae7d425d693a7d964bbf66106f. Please rebase on fresh `develop` branch and use it here instead of write().

I've also fixed several issues with our glibc backtrace implementation: d7971de2b23e3f597cfd5ddcb19c6131fdc2d41c. Not related to this PR, just FYI.